reference/src/layout/unions.md

```diff
@@ -38,3 +28,83 @@ contents are.
 Even if the offsets happen to be all 0, there might still be differences in the
 function call ABI.  If you need to pass unions by-value across an FFI boundary,
 you have to use `#[repr(C)]`.
+
+### Layout of "repr C" unions
+
+The layout of `repr(C)` unions follows the C layout scheme. Per sections
+[6.5.8.5] and [6.7.2.1.16] of the C11 specification, this means that the offset
+of every field is 0. Unsafe code can cast a pointer to the union to a field type
+to obtain a pointer to any field, and vice versa. 
+
+[6.5.8.5]: http://port70.net/~nsz/c/c11/n1570.html#6.5.8p5
+[6.7.2.1.16]: http://port70.net/~nsz/c/c11/n1570.html#6.7.2.1p16
+
+#### Padding
+
+Since all fields are at offset 0, `repr(C)` unions do not have padding before
+their fields. They can, however, have trailing padding, to make sure the size is
+a multiple of the alignment:
+
+```rust
+# use std::mem::{size_of, align_of};
+#[repr(C, align(2))]
+union U { x: u8 }
+# fn main() {
+// The repr(align) attribute raises the alignment requirement of U to 2
+assert_eq!(align_of::<U>(), 2);
+// This introduces trailing padding, raising the union size to 2
+assert_eq!(size_of::<U>(), 2);
+# }
+``**
+
+**Note**: there is no room between fields for padding, so `repr(C)` unions can
+only have trailing padding.
+
+The bit `i` of a `repr(C)` union is a padding bit if the bit `i` of each of its
+fields is a padding bit or trailing padding. That is:
+
+```rust
+#[repr(C)]
+union U { x: (u8, u16) }
+```
+
+The byte at offset 1 of `U` is a padding byte.
```

Comment:
I think the bit starting at "The bit `i`" and ending here should be dropped. This defines what "the padding bytes of a union" are, but we do not even have an agreed-upon definition for "the padding bytes of a type", of which this would have to be an instance.

Reply:
We finally get to the main change being made here 😆 

In this definition, "padding" is meant in the "C" language sense. AFAICT, we don't have much freedom here. Once we have a better definition of what padding is somewhere else, we should try to reconcile that with this one. But for `repr(C)` unions, interior field padding is not copied on move / copy / .. of the union, and in C one would say that this interior field padding is padding of the union itself.

Reply:
I don't know what "padding in the C language sense" is or means.

My guess is you mean "not preserved on copies". I think that's out of scope for this document -- this is about *layout*, not *value representation*, and it is the latter that determines things like what gets preserved on a copy.

Reply:
Makes sense. I'll remove this, but we do document the value representation of some types in the layout section (e.g. that a bool is 0 or 1). We need to discuss the value representation of unions at some point.

Reply:
> but we do document the value representation of some types in the layout section (e.g. that a bool is 0 or 1)

Fair. But for bool it is very simple. ;)

>  We need to discuss the value representation of unions at some point.

Absolutely.